src/sync/impl/sync_file.cpp

```diff
@@ -0,0 +1,262 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync_file.hpp"
+
+#include <realm/util/file.hpp>
+
+using File = realm::util::File;
+
+namespace realm {
+
+static uint8_t value_of_hex_digit(char hex_digit)
+{
+    if (hex_digit >= '0' && hex_digit <= '9') {
+        return hex_digit - '0';
+    } else if (hex_digit >= 'A' && hex_digit <= 'F') {
+        return 10 + hex_digit - 'A';
+    } else if (hex_digit >= 'a' && hex_digit <= 'f') {
+        return 10 + hex_digit - 'a';
+    } else {
+        throw std::invalid_argument("Cannot get the value of a character that isn't a hex digit.");
+    }
+}
+
+static bool character_is_unreserved(char character)
+{
+    bool is_capital_letter = (character >= 'A' && character <= 'Z');
+    bool is_lowercase_letter = (character >= 'a' && character <= 'z');
+    bool is_number = (character >= '0' && character <= '9');
+    bool is_allowed_symbol = (character == '-' || character == '_');
+    return is_capital_letter || is_lowercase_letter || is_number || is_allowed_symbol;
+}
+
+static char decoded_char_for(const std::string& percent_encoding, size_t index)
+{
+    if (index+2 >= percent_encoding.length()) {
+        throw std::invalid_argument("Malformed string: not enough characters after '%' before end of string.");
+    }
+    REALM_ASSERT(percent_encoding[index] == '%');
+    return (16*value_of_hex_digit(percent_encoding[index + 1])) + value_of_hex_digit(percent_encoding[index + 2]);
+}
```

Comment:
Suggest putting these functions in an anonymous namespace instead of using `static`, which is more consistent with the rest of our codebase. There's no semantic difference, but `static` is a little C-like. :-P